File: iOS/Spaces V1.5/Services/AnnouncementStore.swift
import Combine
import FirebaseCore
import FirebaseFirestore
import Foundation

@MainActor
final class AnnouncementStore: ObservableObject {
    static let shared = AnnouncementStore()

    @Published private var announcementsBySpace: [String: [SpaceAnnouncement]] = [:]
    @Published private(set) var lastErrorMessage: String?

    private var listeners: [String: ListenerRegistration] = [:]
    private var spacesByID: [String: Space] = [:]

    private init() {}

    func startListening(in space: Space) {
        spacesByID[space.id] = space
        guard listeners[space.id] == nil else { return }
        guard FirebaseApp.app() != nil else {
            ensureSeeded(space)
            return
        }

        let collection = Firestore.firestore()
            .collection("spaces")
            .document(space.id)
            .collection("announcements")

        listeners[space.id] = collection.addSnapshotListener { [weak self] snapshot, error in
            Task { @MainActor [weak self] in
                guard let self else { return }
                if let error {
                    self.lastErrorMessage = error.localizedDescription
                    return
                }

                self.announcementsBySpace[space.id] = snapshot?.documents.compactMap {
                    self.mapAnnouncement(document: $0, spaceID: space.id)
                } ?? []
                self.lastErrorMessage = nil
            }
        }
    }

    func announcements(in space: Space) -> [SpaceAnnouncement] {
        ensureSeeded(space)
        return (announcementsBySpace[space.id] ?? [])
            .filter { !$0.isExpired }
            .sorted {
                if $0.isPinned != $1.isPinned {
                    return $0.isPinned
                }
                return $0.createdAt > $1.createdAt
            }
    }

    func announcement(id: String, in space: Space) -> SpaceAnnouncement? {
        ensureSeeded(space)
        return announcementsBySpace[space.id]?.first(where: { $0.id == id })
    }

    func save(_ announcement: SpaceAnnouncement) {
        var values = announcementsBySpace[announcement.spaceID] ?? []
        let isNew = !values.contains { $0.id == announcement.id }
        if let index = values.firstIndex(where: { $0.id == announcement.id }) {
            values[index] = announcement
        } else {
            values.append(announcement)
        }
        announcementsBySpace[announcement.spaceID] = values
        persist(announcement)
        if let space = spacesByID[announcement.spaceID] {
            Task {
                let service = SpaceService()
                if isNew {
                    await service.recordModuleActivity(
                        type: .announcementCreated,
                        in: space,
                        title: "posted an Announcement",
                        subtitle: announcement.title,
                        targetID: announcement.id,
                        targetType: .announcements
                    )
                }
                if !isNew {
                    let mentionedIDs = await service.mentionedMemberIDs(in: announcement.body, space: space)
                    await service.recordTargetedModuleNotification(
                        recipientIDs: mentionedIDs,
                        type: .announcement,
                        in: space,
                        title: "mentioned you in an Announcement",
                        subtitle: announcement.title,
                        targetID: announcement.id,
                        targetType: .announcements
                    )
                }
            }
        }
    }

    func delete(id: String, in space: Space) {
        announcementsBySpace[space.id]?.removeAll { $0.id == id }
        guard FirebaseApp.app() != nil else { return }
        Firestore.firestore()
            .collection("spaces")
            .document(space.id)
            .collection("announcements")
            .document(id)
            .delete { [weak self] error in
                guard let error else { return }
                Task { @MainActor [weak self] in
                    self?.lastErrorMessage = error.localizedDescription
                }
            }
    }

    func toggleReaction(emoji: String, announcementID: String, space: Space, userID: String) {
        update(announcementID: announcementID, space: space) { announcement in
            if let index = announcement.reactions.firstIndex(where: { $0.emoji == emoji }) {
                if announcement.reactions[index].userIDs.contains(userID) {
                    announcement.reactions[index].userIDs.remove(userID)
                } else {
                    announcement.reactions[index].userIDs.insert(userID)
                }
                announcement.reactions.removeAll { $0.userIDs.isEmpty }
            } else {
                announcement.reactions.append(.init(emoji: emoji, userIDs: [userID]))
            }
            announcement.updatedAt = Date()
        }
        persistCurrent(announcementID: announcementID, space: space)
    }

    func addComment(_ comment: AnnouncementComment, announcementID: String, space: Space) {
        update(announcementID: announcementID, space: space) { announcement in
            guard announcement.commentsEnabled else { return }
            announcement.comments.append(comment)
            announcement.updatedAt = Date()
        }
        persistCurrent(announcementID: announcementID, space: space)
        Task {
            let service = SpaceService()
            let mentionedIDs = await service.mentionedMemberIDs(in: comment.body, space: space)
            await service.recordTargetedModuleNotification(
                recipientIDs: Array(Set(mentionedIDs + (announcement(id: announcementID, in: space).map { [$0.authorID] } ?? []))),
                type: .announcement,
                in: space,
                title: "commented on your Announcement",
                subtitle: announcement(id: announcementID, in: space)?.title,
                targetID: announcementID,
                targetType: .announcements
            )
        }
    }

    private func update(
        announcementID: String,
        space: Space,
        mutation: (inout SpaceAnnouncement) -> Void
    ) {
        ensureSeeded(space)
        guard var values = announcementsBySpace[space.id],
              let index = values.firstIndex(where: { $0.id == announcementID }) else {
            return
        }
        mutation(&values[index])
        announcementsBySpace[space.id] = values
    }

    private func ensureSeeded(_ space: Space) {
        guard announcementsBySpace[space.id] == nil else { return }
        announcementsBySpace[space.id] = []
    }

    private func persistCurrent(announcementID: String, space: Space) {
        guard let announcement = announcementsBySpace[space.id]?.first(where: { $0.id == announcementID }) else {
            return
        }
        persist(announcement)
    }

    private func persist(_ announcement: SpaceAnnouncement) {
        guard FirebaseApp.app() != nil else { return }

        Firestore.firestore()
            .collection("spaces")
            .document(announcement.spaceID)
            .collection("announcements")
            .document(announcement.id)
            .setData(documentData(for: announcement), merge: true) { [weak self] error in
                guard let error else { return }
                Task { @MainActor [weak self] in
                    self?.lastErrorMessage = error.localizedDescription
                }
            }
    }

    private func documentData(for announcement: SpaceAnnouncement) -> [String: Any] {
        [
            "spaceId": announcement.spaceID,
            "title": announcement.title,
            "body": announcement.body,
            "authorId": announcement.authorID,
            "authorName": announcement.authorName,
            "createdAt": Timestamp(date: announcement.createdAt),
            "updatedAt": Timestamp(date: announcement.updatedAt),
            "isPinned": announcement.isPinned,
            "expiresAt": announcement.expiresAt.map(Timestamp.init(date:)) ?? NSNull(),
            "commentsEnabled": announcement.commentsEnabled,
            "attachments": announcement.attachments.map {
                [
                    "id": $0.id,
                    "kind": $0.kind.rawValue,
                    "title": $0.title,
                    "url": $0.urlString.map { $0 as Any } ?? NSNull(),
                    "storagePath": $0.storagePath.map { $0 as Any } ?? NSNull(),
                    "nonce": $0.nonce.map { $0 as Any } ?? NSNull(),
                    "mimeType": $0.mimeType.map { $0 as Any } ?? NSNull(),
                    "fileSize": $0.fileSize.map { $0 as Any } ?? NSNull(),
                    "uploadedBy": $0.uploadedBy.map { $0 as Any } ?? NSNull()
                ]
            },
            "references": announcement.references.map {
                [
                    "id": $0.id,
                    "kind": $0.kind.rawValue,
                    "targetId": $0.targetID,
                    "title": $0.title,
                    "subtitle": $0.subtitle.map { $0 as Any } ?? NSNull()
                ]
            },
            "reactions": announcement.reactions.map {
                [
                    "emoji": $0.emoji,
                    "userIds": Array($0.userIDs)
                ]
            },
            "comments": announcement.comments.map {
                [
                    "id": $0.id,
                    "authorId": $0.authorID,
                    "authorName": $0.authorName,
                    "body": $0.body,
                    "createdAt": Timestamp(date: $0.createdAt)
                ]
            }
        ]
    }

    private func mapAnnouncement(document: QueryDocumentSnapshot, spaceID: String) -> SpaceAnnouncement? {
        let data = document.data()
        guard
            let title = data["title"] as? String,
            let body = data["body"] as? String,
            let authorID = data["authorId"] as? String,
            let authorName = data["authorName"] as? String
        else {
            return nil
        }

        let attachments = (data["attachments"] as? [[String: Any]] ?? []).compactMap { value -> AnnouncementAttachment? in
            guard
                let kindValue = value["kind"] as? String,
                let kind = AnnouncementAttachmentKind(rawValue: kindValue),
                let title = value["title"] as? String
            else {
                return nil
            }
            return AnnouncementAttachment(
                id: value["id"] as? String ?? UUID().uuidString,
                kind: kind,
                title: title,
                urlString: value["url"] as? String,
                storagePath: value["storagePath"] as? String,
                nonce: value["nonce"] as? String,
                mimeType: value["mimeType"] as? String,
                fileSize: value["fileSize"] as? Int,
                uploadedBy: value["uploadedBy"] as? String
            )
        }.uniquedByID()

        let references = (data["references"] as? [[String: Any]] ?? []).compactMap { value -> AnnouncementReference? in
            guard
                let kindValue = value["kind"] as? String,
                let kind = AnnouncementReferenceKind(rawValue: kindValue),
                let targetID = value["targetId"] as? String,
                let title = value["title"] as? String
            else {
                return nil
            }
            return AnnouncementReference(
                id: value["id"] as? String ?? UUID().uuidString,
                kind: kind,
                targetID: targetID,
                title: title,
                subtitle: value["subtitle"] as? String
            )
        }.uniquedByID()

        let reactions = (data["reactions"] as? [[String: Any]] ?? []).compactMap { value -> AnnouncementReaction? in
            guard let emoji = value["emoji"] as? String else { return nil }
            return AnnouncementReaction(
                emoji: emoji,
                userIDs: Set(value["userIds"] as? [String] ?? [])
            )
        }.uniquedByID()

        let comments = (data["comments"] as? [[String: Any]] ?? []).compactMap { value -> AnnouncementComment? in
            guard
                let authorID = value["authorId"] as? String,
                let authorName = value["authorName"] as? String,
                let body = value["body"] as? String
            else {
                return nil
            }
            return AnnouncementComment(
                id: value["id"] as? String ?? UUID().uuidString,
                authorID: authorID,
                authorName: authorName,
                body: body,
                createdAt: (value["createdAt"] as? Timestamp)?.dateValue() ?? Date()
            )
        }.uniquedByID()

        return SpaceAnnouncement(
            id: document.documentID,
            spaceID: data["spaceId"] as? String ?? spaceID,
            title: title,
            body: body,
            authorID: authorID,
            authorName: authorName,
            createdAt: (data["createdAt"] as? Timestamp)?.dateValue() ?? Date(),
            updatedAt: (data["updatedAt"] as? Timestamp)?.dateValue() ?? Date(),
            isPinned: data["isPinned"] as? Bool ?? false,
            expiresAt: (data["expiresAt"] as? Timestamp)?.dateValue(),
            commentsEnabled: data["commentsEnabled"] as? Bool ?? true,
            attachments: attachments,
            references: references,
            reactions: reactions,
            comments: comments
        )
    }

}

File: Android V1.6/app/src/main/java/com/arcinteractive/spaces/data/spaces/AnnouncementStore.kt
package com.arcinteractive.spaces.data.spaces

import android.content.Context
import com.arcinteractive.spaces.data.model.AnnouncementAttachment
import com.arcinteractive.spaces.data.model.AnnouncementAttachmentKind
import com.arcinteractive.spaces.data.model.AnnouncementComment
import com.arcinteractive.spaces.data.model.AnnouncementReaction
import com.arcinteractive.spaces.data.model.AnnouncementReference
import com.arcinteractive.spaces.data.model.AnnouncementReferenceKind
import com.arcinteractive.spaces.data.model.Space
import com.arcinteractive.spaces.data.model.SpaceAnnouncement
import com.arcinteractive.spaces.data.media.EncryptedMediaService
import com.google.firebase.FirebaseApp
import com.google.firebase.Timestamp
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ListenerRegistration
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.Date
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import com.arcinteractive.spaces.data.model.ActivityTargetType
import com.arcinteractive.spaces.data.model.ActivityType

object AnnouncementStore {
    private val _announcements = MutableStateFlow<Map<String, List<SpaceAnnouncement>>>(emptyMap())
    val announcements: StateFlow<Map<String, List<SpaceAnnouncement>>> = _announcements.asStateFlow()

    private val _lastErrorMessage = MutableStateFlow<String?>(null)
    val lastErrorMessage: StateFlow<String?> = _lastErrorMessage.asStateFlow()

    private val listeners = mutableMapOf<String, ListenerRegistration>()
    private var firestore: FirebaseFirestore? = null
    private var applicationContext: Context? = null
    private val spacesById = mutableMapOf<String, Space>()
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    fun startListening(context: Context, space: Space) {
        applicationContext = context.applicationContext
        spacesById[space.id] = space
        if (listeners.containsKey(space.id)) return
        val database = firestore(context) ?: run {
            ensureSeeded(space)
            return
        }

        if (!_announcements.value.containsKey(space.id)) {
            _announcements.value = _announcements.value + (space.id to emptyList())
        }
        listeners[space.id] = database.collection("spaces")
            .document(space.id)
            .collection("announcements")
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    _lastErrorMessage.value = error.localizedMessage ?: "Announcements could not be loaded."
                    return@addSnapshotListener
                }
                val values = snapshot?.documents.orEmpty().mapNotNull { mapAnnouncement(it, space.id) }
                _announcements.value = _announcements.value + (space.id to values)
                _lastErrorMessage.value = null
            }
    }

    fun clearError() {
        _lastErrorMessage.value = null
    }

    suspend fun uploadAttachment(
        context: Context,
        spaceId: String,
        userId: String,
        fileName: String,
        mimeType: String,
        bytes: ByteArray
    ): AnnouncementAttachment {
        val id = java.util.UUID.randomUUID().toString()
        val metadata = EncryptedMediaService().uploadFile(
            context = context,
            spaceId = spaceId,
            storagePath = "spaces/$spaceId/announcements/$id",
            originalBytes = bytes,
            mimeType = mimeType,
            uploadedBy = userId
        ).metadata
        val kind = when {
            mimeType.startsWith("image/") -> AnnouncementAttachmentKind.Image
            mimeType.startsWith("video/") -> AnnouncementAttachmentKind.Video
            else -> AnnouncementAttachmentKind.File
        }
        return AnnouncementAttachment(
            id = id,
            kind = kind,
            title = fileName,
            storagePath = metadata.storagePath,
            nonce = metadata.nonce,
            mimeType = metadata.mimeType,
            fileSize = metadata.fileSize,
            uploadedBy = metadata.uploadedBy
        )
    }

    fun forSpace(space: Space): List<SpaceAnnouncement> {
        if (!_announcements.value.containsKey(space.id)) ensureSeeded(space)
        return _announcements.value[space.id].orEmpty()
            .filterNot { it.isExpired }
            .sortedWith(compareByDescending<SpaceAnnouncement> { it.isPinned }.thenByDescending { it.createdAt })
    }

    fun save(announcement: SpaceAnnouncement) {
        val values = _announcements.value[announcement.spaceId].orEmpty().toMutableList()
        val index = values.indexOfFirst { it.id == announcement.id }
        val isNew = index < 0
        if (index >= 0) values[index] = announcement else values += announcement
        _announcements.value = _announcements.value + (announcement.spaceId to values)
        persist(announcement)
        val context = applicationContext
        val space = spacesById[announcement.spaceId]
        if (context != null && space != null) scope.launch {
            val service = SpaceService()
            if (isNew) {
                service.recordModuleActivity(
                    context, ActivityType.AnnouncementCreated, space, "posted an Announcement",
                    announcement.title, announcement.id, ActivityTargetType.Announcements
                )
            }
            if (!isNew) {
                val recipients = service.mentionedMemberIds(context, announcement.body, space)
                service.recordTargetedModuleNotification(
                    context, recipients, "announcement", space, "mentioned you in an Announcement",
                    announcement.title, announcement.id, ActivityTargetType.Announcements
                )
            }
        }
    }

    fun delete(spaceId: String, announcementId: String) {
        _announcements.value = _announcements.value + (
            spaceId to _announcements.value[spaceId].orEmpty().filterNot { it.id == announcementId }
        )
        firestore?.collection("spaces")?.document(spaceId)?.collection("announcements")
            ?.document(announcementId)?.delete()
            ?.addOnFailureListener { _lastErrorMessage.value = it.localizedMessage }
    }

    fun toggleReaction(spaceId: String, announcementId: String, emoji: String, userId: String) {
        update(spaceId, announcementId) { announcement ->
            val reactions = announcement.reactions.toMutableList()
            val index = reactions.indexOfFirst { it.emoji == emoji }
            if (index >= 0) {
                val users = reactions[index].userIds.toMutableSet()
                if (!users.add(userId)) users.remove(userId)
                if (users.isEmpty()) reactions.removeAt(index) else reactions[index] = reactions[index].copy(userIds = users)
            } else {
                reactions += AnnouncementReaction(emoji, setOf(userId))
            }
            announcement.copy(reactions = reactions, updatedAt = Date())
        }
    }

    fun addComment(spaceId: String, announcementId: String, comment: AnnouncementComment) {
        val creatorId = _announcements.value[spaceId].orEmpty().firstOrNull { it.id == announcementId }?.authorId
        val title = _announcements.value[spaceId].orEmpty().firstOrNull { it.id == announcementId }?.title
        update(spaceId, announcementId) { announcement ->
            if (!announcement.commentsEnabled) announcement
            else announcement.copy(comments = announcement.comments + comment, updatedAt = Date())
        }
        val context = applicationContext
        val space = spacesById[spaceId]
        if (context != null && space != null && creatorId != null) scope.launch {
            val service = SpaceService()
            val recipients = (service.mentionedMemberIds(context, comment.body, space) + creatorId).distinct()
            SpaceService().recordTargetedModuleNotification(
                context, recipients, "announcement", space, "commented on your Announcement",
                title, announcementId, ActivityTargetType.Announcements
            )
        }
    }

    private fun update(spaceId: String, announcementId: String, transform: (SpaceAnnouncement) -> SpaceAnnouncement) {
        var changed: SpaceAnnouncement? = null
        val values = _announcements.value[spaceId].orEmpty().map {
            if (it.id == announcementId) transform(it).also { updated -> changed = updated } else it
        }
        _announcements.value = _announcements.value + (spaceId to values)
        changed?.let(::persist)
    }

    private fun persist(announcement: SpaceAnnouncement) {
        firestore?.collection("spaces")?.document(announcement.spaceId)?.collection("announcements")
            ?.document(announcement.id)?.set(documentData(announcement))
            ?.addOnFailureListener { _lastErrorMessage.value = it.localizedMessage }
    }

    private fun firestore(context: Context): FirebaseFirestore? {
        firestore?.let { return it }
        if (FirebaseApp.getApps(context.applicationContext).isEmpty()) return null
        return FirebaseFirestore.getInstance().also { firestore = it }
    }

    private fun documentData(value: SpaceAnnouncement): Map<String, Any?> = mapOf(
        "spaceId" to value.spaceId,
        "title" to value.title,
        "body" to value.body,
        "authorId" to value.authorId,
        "authorName" to value.authorName,
        "createdAt" to Timestamp(value.createdAt),
        "updatedAt" to Timestamp(value.updatedAt),
        "isPinned" to value.isPinned,
        "expiresAt" to value.expiresAt?.let(::Timestamp),
        "commentsEnabled" to value.commentsEnabled,
        "attachments" to value.attachments.map {
            mapOf(
                "id" to it.id, "kind" to it.kind.name.lowercase(), "title" to it.title, "url" to it.url,
                "storagePath" to it.storagePath, "nonce" to it.nonce, "mimeType" to it.mimeType,
                "fileSize" to it.fileSize, "uploadedBy" to it.uploadedBy
            )
        },
        "references" to value.references.map {
            mapOf(
                "id" to it.id,
                "kind" to if (it.kind == AnnouncementReferenceKind.ListItem) "list" else it.kind.name.lowercase(),
                "targetId" to it.targetId,
                "title" to it.title,
                "subtitle" to it.subtitle
            )
        },
        "reactions" to value.reactions.map { mapOf("emoji" to it.emoji, "userIds" to it.userIds.toList()) },
        "comments" to value.comments.map {
            mapOf(
                "id" to it.id,
                "authorId" to it.authorId,
                "authorName" to it.authorName,
                "body" to it.body,
                "createdAt" to Timestamp(it.createdAt)
            )
        }
    )

    private fun mapAnnouncement(document: DocumentSnapshot, fallbackSpaceId: String): SpaceAnnouncement? {
        val data = document.data ?: return null
        val attachments = data.listOfMaps("attachments").mapNotNull {
            val kind = enumValues<AnnouncementAttachmentKind>().firstOrNull { value ->
                value.name.equals(it["kind"] as? String, ignoreCase = true)
            } ?: return@mapNotNull null
            AnnouncementAttachment(
                id = it["id"] as? String ?: return@mapNotNull null,
                kind = kind,
                title = it["title"] as? String ?: "",
                url = it["url"] as? String,
                storagePath = it["storagePath"] as? String,
                nonce = it["nonce"] as? String,
                mimeType = it["mimeType"] as? String,
                fileSize = (it["fileSize"] as? Number)?.toLong(),
                uploadedBy = it["uploadedBy"] as? String
            )
        }
        val references = data.listOfMaps("references").mapNotNull {
            val rawKind = it["kind"] as? String ?: return@mapNotNull null
            val kind = if (rawKind.equals("list", true)) AnnouncementReferenceKind.ListItem
            else enumValues<AnnouncementReferenceKind>().firstOrNull { value -> value.name.equals(rawKind, true) }
                ?: return@mapNotNull null
            AnnouncementReference(
                id = it["id"] as? String ?: return@mapNotNull null,
                kind = kind,
                targetId = it["targetId"] as? String ?: return@mapNotNull null,
                title = it["title"] as? String ?: "",
                subtitle = it["subtitle"] as? String
            )
        }
        return SpaceAnnouncement(
            id = document.id,
            spaceId = data["spaceId"] as? String ?: fallbackSpaceId,
            title = data["title"] as? String ?: return null,
            body = data["body"] as? String ?: "",
            authorId = data["authorId"] as? String ?: "",
            authorName = data["authorName"] as? String ?: "Unknown",
            createdAt = data.date("createdAt") ?: Date(),
            updatedAt = data.date("updatedAt") ?: Date(),
            isPinned = data["isPinned"] as? Boolean ?: false,
            expiresAt = data.date("expiresAt"),
            commentsEnabled = data["commentsEnabled"] as? Boolean ?: true,
            attachments = attachments,
            references = references,
            reactions = data.listOfMaps("reactions").mapNotNull {
                val emoji = it["emoji"] as? String ?: return@mapNotNull null
                AnnouncementReaction(emoji, (it["userIds"] as? List<*>)?.filterIsInstance<String>()?.toSet().orEmpty())
            },
            comments = data.listOfMaps("comments").mapNotNull {
                AnnouncementComment(
                    id = it["id"] as? String ?: return@mapNotNull null,
                    authorId = it["authorId"] as? String ?: "",
                    authorName = it["authorName"] as? String ?: "Unknown",
                    body = it["body"] as? String ?: "",
                    createdAt = it.date("createdAt") ?: Date()
                )
            }
        )
    }

    @Suppress("UNCHECKED_CAST")
    private fun Map<String, Any>.listOfMaps(key: String): List<Map<String, Any>> =
        (this[key] as? List<*>)?.mapNotNull { it as? Map<String, Any> }.orEmpty()

    private fun Map<String, Any>.date(key: String): Date? = when (val value = this[key]) {
        is Timestamp -> value.toDate()
        is Date -> value
        else -> null
    }

    private fun ensureSeeded(space: Space) {
        if (_announcements.value.containsKey(space.id)) return
        _announcements.value = _announcements.value + (space.id to emptyList())
    }
}
